types/react/index.d.ts

```diff
@@ -121,7 +122,7 @@ declare namespace React {
         : C extends ((props: any, context?: any) => ReactElement | null)
         ? undefined
         : C extends keyof JSX.IntrinsicElements
-        ? JSX.IntrinsicElements[C] extends DOMAttributes<infer E>
+        ? JSX.IntrinsicElements[C] extends JSX.IntrinsicClassAttributes<infer E>
```

Comment:
I haven't really seen any usage of this in the wild, can you point to an example of a custom renderer that augments the `IntrinsicElements` interface? This seems kind of weird.

Reply:
https://www.typescriptlang.org/docs/handbook/jsx.html#intrinsic-elements

If I'm right about understanding the JSX namespace, it shouldn't be written to depend on a particular host implementation.

The same behavior must be guaranteed for react-three-fiber or react-pixi. However, ElementType infers the ref of react-three-fiber host elements to `never`.

https://github.com/react-spring/react-three-fiber/blob/master/src/three-types.ts

Reply:
React-three-fiber's IntrinsicElements cannot be inferred from IntrinsicClassAttributes, so it doesn't work anyway. But it can be fixed.

Reply:
react-three-fiber's types augments the intrinsic elements by adding lowercase host "tags", but what ref do they support? IE what does the reconciler give you at runtime when you do this?
```ts
const cameraRef = createRef()
const myCamera = <camera ref={cameraRef} />
```

Looking at its types, it is telling me that it would give me a `ReactNode` which seems inaccurate.

Reply:
I'm also noticing that React fiber's types don't augment `JSX.IntrinsicClassAttributes` so I don't believe that the changes in the PR would help 🤔

Reply:
> react-three-fiber's types augments the intrinsic elements by adding lowercase host "tags", but what ref do they support? i.e. what does the reconciler give you at runtime when you do this?

> I'm also noticing that React fiber's types don't augment JSX.IntrinsicClassAttributes so I don't believe that the changes in the PR would help 🤔

>  so it doesn't work anyway. But it can be fixed.

As you said, It would not help for right now.

However, custom renderers like react-three-fiber can declare JSX.IntrinsicClassAttributes to make the possibility to interop with React type definitions.

Other renderers should also be available, ink, react-pixi, react-native, etc... by defining attributes like React's DetailedHTMLAttributes. (I think they should also move to react-dom BTW)

Reply:
The current problem is that the definition of react is only written for react-dom we should avoid.

Reply:
My point is more that other reconcilers that use their own interface for `IntrinsicAttributes` (barring the current problem of react typedefs using a global `JSX` namespace) to add host elements is that it won't be able to infer via `IntrinsicClassAttributes`.

The generic parameter `T` in `JSX.IntrinsicClassAttributes` is neither an inference site nor a constraint check, it's actually passed in by the compiler *after* it checks for `ElementClass` compatibility.

It goes something like...
- Given an opaque JSX element expression:
- Look at the tag, if it's uppercase:
  - Check if the symbol associated with the tag is of the signature `() => JSX.Element`, if so we're okay
  - Otherwise, check if the symbol associated with the tag is of the signature `new () => JSX.ElementClass`, if so then we're okay
- If it's lowercase:
  - Check if the string associated is a key of `JSX.IntrinsicElements`, if so we're okay

Once those checks pass, then we go into props checking which is far more complicated, but the gist of it is that given the props you pass, it checks if it's compatible with the following type:
```ts
// if the tag's symbol is not a class (lowercase or uppercase with () => JSX.Element signature)
type AllowedAttributesNonClass = JSX.LibraryManagedAttributes<TagSymbol, PropsYouPass> & JSX.IntrinsicAttributes
// if the tag's symbol is a class
type AllowedAttributesClass = AllowedAttributesNonClass & JSX.IntrinsicClassAttributes<InstanceType<TagSymbol>>
```

When we're talking about `ElementRef`, we don't consider any of the above `JSX` types at all, so `IntrinsicClassAttributes` doesn't actually come into play, nor `IntrinsicAttributes`.

Reply:
*If* another reconciler wants to be able to provide refs accurately, they'd need to do something like:
```ts
// my-reconciler.d.ts
declare global {
  declare namespace JSX {
    interface IntrinsicElements {
      myCustomHostElement: OtherProps & { ref?: ThisIsMyInstanceType }
    }
  }
}
```

Reply:
SECOND EDIT: okay, I think I know what my reservations are now. `IntrinsicClassAttributes` is being used by the current HTML elements' interface, but it's not necessary. I think a more generic and broad way of implementing this is something like:
```ts
type ElementRef<C> = C extends keyof JSX.IntrinsicElements 
  ? JSX.IntrinsicElements[C] extends { ref?: Ref<infer T> } 
    ? T : unknown 
  : TheRestOfTheStuff<C>
```

Reply:
@ferdaber In React's current type definition, `IntrinsicClassAttributes<infer E>` is compatible with `DetailedHTMLProps` and also `SVGProps`. But your suggestion is not compatible with `SVGProps`.

Should I fix the `SVGProps` definition to be compatible with `React.RefAttributes` here?

Reply:
What do you mean? Is `ElementRef` as implemented like my above comment not working for SVG tags?

Reply:
@ferdaber Ah, I've got confused.

```ts
type ElementRef_1<C> = C extends keyof JSX.IntrinsicElements 
  ? JSX.IntrinsicElements[C] extends { ref?: Ref<infer T> } 
    ? T : never 
  : TheRestOfTheStuff<C>

type ElementRef_2<C> = C extends keyof JSX.IntrinsicElements 
  ? JSX.IntrinsicElements[C] extends { ref?: RefObject<infer T> } 
    ? T : never 
  : TheRestOfTheStuff<C>

type ElementRef_3<C> = C extends keyof JSX.IntrinsicElements 
  ? JSX.IntrinsicElements[C] extends { ref?: infer R }
    ? R extends Ref<infer E>
      ? E
      : never
    : never 
  : TheRestOfTheStuff<C>

type ElementRef_4<C> = C extends keyof JSX.IntrinsicElements 
  ? JSX.IntrinsicElements[C] extends { ref?: infer R }
    ? R extends RefObject<infer E>
      ? E
      : never
    : never 
  : TheRestOfTheStuff<C>
```

Only `ElementRef_4<C>` works as I expected, and others return `never`... I didn't understand why

And using `unknown` never works because If any branch can be unknown, all results will be unknown in TypeScript.

Reply:
Submitted the change as you suggested. Thank you @ferdaber